Deployment preflight errors due to bad content GUID

marcosnav opened the pull request:
## Intent

**Main intent:**
Better error messaging on deployment errors due to bad content ID on deployment toml

**Common errors handled by this PR**
- The GUID does not match the content or does not exist - 404
<img width="1441" alt="Screen Shot 2024-10-03 at 15 03 51" src="https://github.com/user-attachments/assets/eb317674-55cb-49a8-9c1c-2c68afb81b58">

- The user does not have permissions anymore to the content - 403
<img width="1343" alt="Screen Shot 2024-10-03 at 15 09 55" src="https://github.com/user-attachments/assets/6bb74f9c-f0b5-4e2b-b86c-823e2a4dc834">

- Content is locked
<img width="1441" alt="Screen Shot 2024-10-03 at 15 03 31" src="https://github.com/user-attachments/assets/b4357b21-5b99-4dac-a427-1af5ec61f9ef">

- Unknown error
<img width="1537" alt="Screen Shot 2024-10-03 at 15 56 55" src="https://github.com/user-attachments/assets/0382171e-02c1-48b2-b118-2fcfdfb28f6b">

Fixes #2287 

## Type of Change

Adding an extra step on the backend when deploying. Before building the bundle and doing any further operations, we will do a preflight check, trying to query the content target to validate that it can be reached, it exists and the user has permissions over it.

Returning an error with enough details to be handled by the extension code.

The changes are taking a similar approach to the http Axios errors but in this case these are messages sent by via SSE, so this is another area of the product where we are now filtering by error codes to provide a more meaningful message.

## Automated Tests

Created and updated unit tests accordingly.

## Directions for Reviewers

- [ ] For an existing content already deployed, change the `id` within the `deployment-*.toml` to:
    - [ ] To a content GUID that you don't have permissions to (or use a new credential without enough permissions to that content)
    - [ ] To a GUID for some content that does not exist

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply: -->
<!--- If you need clarification on any of these, feel free to ask. We're here to help! -->

- [ ] I have updated [CHANGELOG.md](../CHANGELOG.md) to cover notable changes.

marcosnav commented on extensions/vscode/src/api/types/events.ts:
```diff
@@ -298,10 +300,11 @@ export function getLocalId(arg: EventStreamMessage) {
   time: string;
-  data: Record<string, string>;
+  data: T;
+  errCode?: ErrorCode;
```
Allowing backend error codes to be passed to SSE messages too

marcosnav commented on extensions/vscode/src/eventErrors.ts:
```diff
@@ -0,0 +1,46 @@
+// Copyright (C) 2024 by Posit Software, PBC.
```
Similar to HTTP axios errors, creating some tooling to filter and handle incoming SSE error messages that have an error code

marcosnav commented on internal/clients/connect/client_connect.go:
```diff
@@ -179,6 +179,11 @@ type connectGetContentDTO struct {
 	// Owner        *ownerOutputDTO   `json:"owner,omitempty"`
 }
 
+func (c *ConnectClient) ContentDetails(contentID types.ContentID, body *ConnectContent, log logging.Logger) error {
```
New client method to pull content data, helps to validate the user has access to it.

marcosnav commented on internal/clients/http_client/http_client.go:
```diff
@@ -241,3 +241,12 @@ func NewHTTPClientForAccount(account *accounts.Account, timeout time.Duration, l
 	}, nil
 }
+
+func IsHTTPAgentErrorStatusOf(err error, status int) (*types.AgentError, bool) {
```
Handy func to check if an error that is bubbling up is an AgentError of certain http status

marcosnav commented on internal/events/events.go:
```diff
@@ -17,13 +17,13 @@ type EventData = types.ErrorData
+	Time    time.Time
+	Type    EventType
+	Data    EventData
+	ErrCode ErrorCode
```
Include the error code on the event payload sent to the extension client

marcosnav commented on internal/publish/publish_test.go:
```diff
@@ -108,119 +123,152 @@ func (s *PublishSuite) TestNewFromState() {
 
 func (s *PublishSuite) TestPublishWithClientNewSuccess() {
-	s.publishWithClient(nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
+	s.publishWithClient(nil, &publishErrsMock{}, nil)
```
Changed a bit the signature here, passing a lot of `nil`s in a certain order was growing too much

dotNomad replied:
Much better!

jonkeane commented on internal/types/error.go:
```diff
@@ -16,6 +16,9 @@ const (
 	ErrorUnknownTOMLKey               ErrorCode = "unknownTOMLKey"
 	ErrorInvalidConfigFiles           ErrorCode = "invalidConfigFiles"
 	ErrorCredentialServiceUnavailable ErrorCode = "credentialsServiceUnavailable"
+	ErrorDeploymentTargetNotFound     ErrorCode = "deploymentTargetNotFound"
+	ErrorDeploymentTargetIsForbidden  ErrorCode = "deploymentTargetIsForbidden"
+	ErrorDeploymentTargetUnreachable  ErrorCode = "deploymentTargetUnreachable"
```
Do we need three unique `deploymentTarget` error types here? Or is it sufficient to pass this information within the error itself? The lines of code are free, but the more types we have the more names we have to keep track of + write in all the places, etc.

sagerb replied:
The concept was that these were three types of errors, which we needed to pass through within the error, so that the UX could identify them and place them into a specific error type (to be handled according per the type). Having them listed in one place allows us to keep them synchronized from one place, since we're not sharing code between the agent and the UX. 

I believe this concept still holds and is a good practice, since the error code here is functioning as a const or enum.

dotNomad commented on internal/publish/publish.go:
```diff
@@ -265,6 +267,28 @@ func (p *defaultPublisher) createDeploymentRecord(
 }
 
+// Does the target exist and the user has permissions to it?
+func (p *defaultPublisher) targetPreflightChecks(client connect.APIClient, contentID types.ContentID) error {
```
I noticed that because the error is retrieved from `ContentDetails` which does a `GET` call and these errors have codes plugged in, but nothing else, the error we write to the deployment record `message` is largely un-helpful. The message is just what the API call error has in `message`

Currently that is the error we show for the "Last Deployment Failed" message in the sidebar. So while I get a great message when the piece of content cannot be found in the notification window the sidebar error is quite generic:

![CleanShot 2024-10-02 at 13 00 20@2x](https://github.com/user-attachments/assets/c3673288-21f9-4717-9dec-23daa0f0ba87)

Not sure if we should consider that part of this work, but I'd argue because its a one line addition it should be.

marcosnav replied:
Maybe improvements to this messaging be related to this other issue #2206?

dotNomad replied:
The methodology to set that message is the same, but they are different parts of the publishing process. Ideally we would solve one at a time.

sagerb replied:
I would think if we created some kind of error message factory, which took in the original error (present either in the stream or in the content record file), then we can consistently produce the enhanced message.

dotNomad commented on extensions/vscode/src/eventErrors.test.ts:
```diff
@@ -0,0 +1,95 @@
+    // Message with another error code
+    let streamMsg = mkEventStreamMsg({}, "unknown");
+    let result = isEvtErrTargetNotFound(
+      streamMsg as EventStreamMessageErrorCoded,
```
The `as` keyword always makes me nervous. This is test code so feel free to ignore this, but I would make `mkEventStreamMsg` have a generic so you could do something like

```
let streamMsg = mkEventStreamMsg<EventStreamMessageErrorCoded>({}, "unknown");
```

or similar. That would remove the need for the `as`. You could even change the return type if `errCode` is passed by overloading the definition of the method.

marcosnav replied:
That's a great idea, I'll try that.

dotNomad commented on extensions/vscode/src/eventErrors.ts:
```diff
@@ -0,0 +1,46 @@
+  emsg: EventStreamMessageErrorCoded,
+): string => {
+  if (isEvtErrTargetNotFound(emsg)) {
+    return `Content at ${emsg.data.dashboardUrl} could not be found. Please, verify the content "id" is accurate.`;
```
I don't think this is an error that most users will be able to fix on their own. 

This is because the `id` is inside of the Content Record under `.posit/publish/deployments` which we never prompt the user to open, and is a little tough to find the correct deployment for your configuration if you have multiple / don't know what to look for.

This is a much better error that what we have now, so perhaps what I'm suggesting should be a follow-up, but we should have a way for users to solve this issue.

marcosnav replied:
I think this provides enough info for now, considering it might be very likely to happen when associating content, the context of the interaction feels appropriate. Totally agree that there's room for improvement but user feedback may provide a better direction.

dotNomad commented on extensions/vscode/src/eventErrors.ts:
```diff
@@ -0,0 +1,46 @@
+    return `Content at ${emsg.data.dashboardUrl} could not be found. Please, verify the content "id" is accurate.`;
+  }
+  if (isEvtErrTargetForbidden(emsg)) {
+    return `You don't have enough permissions to deploy to ${emsg.data.dashboardUrl}. Please, verify the credentials in use.`;
```
Should we include something about requesting Collaborator permissions from the owner as another option?

marcosnav replied:
Makes sense, WDYT of?
```
... .You may need to request collaborator permissions or verify the credentials in use
```

dotNomad replied:
That sounds great 👍

dotNomad commented on extensions/vscode/src/eventErrors.ts:
```diff
@@ -0,0 +1,46 @@
+  if (isEvtErrTargetForbidden(emsg)) {
+    return `You don't have enough permissions to deploy to ${emsg.data.dashboardUrl}. Please, verify the credentials in use.`;
+  }
+  const unknownErrMsg = emsg.data.error || emsg.data.message;
+  return `Unknown error: ${unknownErrMsg}`;
+};
```
After a thorough review of this PR I think I understand @jonkeane's point a bit more about the codes for event messages.

These errors don't look to be doing the same amount of construction we expect the frontend to do for other errors (components, links, buttons, etc).

A highlight of this is the locked error. If you lock a piece of content and attempt to deploy you get a pretty good error even though it is unknown:

![CleanShot 2024-10-02 at 13 17 01@2x](https://github.com/user-attachments/assets/e13da7db-9b98-42b0-8f5c-84cdedf834af)

That didn't need a error code, if statements, or really much at all for it to make sense to the user.

Of course we could add a nice link to the content in the notification to adjust the lock setting, and if we want that we need a code to understand which error was passed up.

marcosnav replied:
That was the initial implementation a few weeks ago, the agent sending the error message and the extension only showing it. The argument of having the flexibility of sending only the error code so the UI can do many things is the approach we have now. Some error codes will require some construction on the UI while others will be fairly minimal.

IMO, at this point, consistency should be the key on where do we build this messages to avoid guessing games for each error message.

In this locked content example, I think it is a nice catch, it should have it's error code and be a known error and not show up as an `Unknown error` message.

dotNomad replied:
> IMO, at this point, consistency should be the key on where do we build this messages to avoid guessing games for each error message.

Completely agree. I lean toward consistency. I was trying to voice my understanding of some of the other concerns brought up, and think out loud how I'm reasoning about which to choose. It sounds like that didn't come across that way - sorry about that.

marcosnav replied:
No worries, glad to hear that. Sharing concerns and thinking out loud always helps on getting to common ground and understanding, I appreciate that 👍

sagerb replied:
I will say that I expect we'll take more and more advantage of having the finely defined codes returned from the agent. I'm going to have to implement a richer error type coming back from the agent for deployment failure, so that I can catch when it has occurred due to TLS Certificate Verification failure, and offer the user the link to adjust the setting. Just wanted to add this context into the discussion. I also lean towards consistency here.

dotNomad reviewed (commented):
I had some concerns about the errors here that I think we can bring to a discussion with the team.

marcosnav commented:
@dotNomad @jonkeane Pushed some changes, addressing the feedback and also found a way to have one single error code for the deployment errors, respecting the message as it comes from the agent. Still having the option to have more specific deployment error codes in the future if there's the need.

dotNomad commented on internal/publish/publish.go:
```diff
@@ -265,6 +267,49 @@ func (p *defaultPublisher) createDeploymentRecord(
+	}
+
+	// Verify content is not locked
+	lockedErr := content.LockedError()
```
Nice that you got the locked error in here too 😄

dotNomad commented on extensions/vscode/src/eventErrors.test.ts:
```diff
@@ -0,0 +1,72 @@
+} from "./eventErrors";
+import { ErrorCode } from "./utils/errorTypes";
+
+function mkEventStreamMsg(data: Record<PropertyKey, any>): EventStreamMessage;
+function mkEventStreamMsg(
```
Little bit more complex up here, but it looks so much more usable in the tests themselves. Great work on the typing!

dotNomad reviewed (approved):
Absolutely fantastic work here. Thanks for addressing those bits of feedback 🙇
